lib/alloc.c

```diff
@@ -71,3 +71,23 @@ xreallocarray(void *p, size_t nmemb, size_t size)
 	        log_get_progname(), strerror(errno));
 	exit(13);
 }
+
+char *
+xstrndup(const char *s, size_t max_length) {
+    size_t length = strlen(s);
+    if (length > max_length) {
+        length = max_length;
+    }
+
+    char *p = malloc(length + 1);
+    if (!p) {
+		fprintf(log_get_logfd(), _("%s: %s\n"),
+	    		log_get_progname(), strerror(errno));
+        exit(13);
+    }
+
+    strncpy(p, s, length);
+    p[length] = '\0';
```

Comment:
This is an antipattern that I've blown away from this project.  Please have a look at the functions in `lib/string/` to find something suitable.

Reply:
You can check that there are no direct calls to strncpy(3) at all:

```c
$ grep -rn '\<strncpy *(' lib* src/
lib/string/strncpy.h:18:#define STRNCPY(dst, src)  strncpy(dst, src, NITEMS(dst))
```

(And that macro is used for other very specific purposes.)

Reply:
New approach is applied by direct usage of a new function: is_valid_user_name